File: ios/ChatAiIos/Models/GeminiModels.swift
import Foundation

// MARK: - Gemini API Request & Response Models

public struct GeminiRequest: Codable {
    public struct Content: Codable {
        public struct Part: Codable {
            public let text: String
            public init(text: String) { self.text = text }
        }
        public let role: String?
        public let parts: [Part]
        public init(role: String? = nil, parts: [Part]) {
            self.role = role
            self.parts = parts
        }
    }

    public struct GenerationConfig: Codable {
        public let temperature: Double?
        public let maxOutputTokens: Int?
        public init(temperature: Double? = 0.7, maxOutputTokens: Int? = 2048) {
            self.temperature = temperature
            self.maxOutputTokens = maxOutputTokens
        }
    }

    public let contents: [Content]
    public let generationConfig: GenerationConfig?

    public init(prompt: String, systemPrompt: String? = nil) {
        var parts: [Content.Part] = []
        if let system = systemPrompt, !system.isEmpty {
            parts.append(Content.Part(text: "System: \(system)\n\n"))
        }
        parts.append(Content.Part(text: prompt))
        self.contents = [Content(role: "user", parts: parts)]
        self.generationConfig = GenerationConfig()
    }
}

public struct GeminiResponse: Codable {
    public struct Candidate: Codable {
        public struct Content: Codable {
            public struct Part: Codable {
                public let text: String?
            }
            public let parts: [Part]?
        }
        public let content: Content?
        public let finishReason: String?
    }

    public let candidates: [Candidate]?

    public var replyText: String? {
        return candidates?.first?.content?.parts?.first?.text
    }
}

File: app/src/main/java/com/chatai/adr/model/GeminiModels.kt
package com.chatai.adr.model

import com.google.gson.annotations.SerializedName

data class GeminiRequest(
    @SerializedName("contents")
    val contents: List<Content>,
    @SerializedName("generationConfig")
    val generationConfig: GenerationConfig? = GenerationConfig()
) {
    data class Content(
        @SerializedName("role")
        val role: String? = null,
        @SerializedName("parts")
        val parts: List<Part>
    )

    data class Part(
        @SerializedName("text")
        val text: String
    )

    data class GenerationConfig(
        @SerializedName("temperature")
        val temperature: Double = 0.7,
        @SerializedName("maxOutputTokens")
        val maxOutputTokens: Int = 2048
    )

    companion object {
        fun create(prompt: String, systemPrompt: String? = null): GeminiRequest {
            val parts = mutableListOf<Part>()
            if (!systemPrompt.isNullOrBlank()) {
                parts.add(Part("System: $systemPrompt\n\n"))
            }
            parts.add(Part(prompt))
            return GeminiRequest(listOf(Content(role = "user", parts = parts)))
        }
    }
}

data class GeminiResponse(
    @SerializedName("candidates")
    val candidates: List<Candidate>? = null
) {
    data class Candidate(
        @SerializedName("content")
        val content: Content? = null,
        @SerializedName("finishReason")
        val finishReason: String? = null
    )

    data class Content(
        @SerializedName("parts")
        val parts: List<Part>? = null
    )

    data class Part(
        @SerializedName("text")
        val text: String? = null
    )

    val replyText: String?
        get() = candidates?.firstOrNull()?.content?.parts?.firstOrNull()?.text
}
